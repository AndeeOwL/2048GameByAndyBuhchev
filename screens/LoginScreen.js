import AsyncStorage from "@react-native-async-storage/async-storage";
import { useNavigation } from "@react-navigation/native";
import { useState } from "react";
import { Alert, StyleSheet, Text, View } from "react-native";
import { Colors } from "../components/common/Colors";
import { Fonts } from "../components/common/Fonts";
import LoginButtons from "../components/Login/LoginButtons";
import LoginForm from "../components/Login/LoginForm";
import MainLogo from "../components/MainLogo";
import GameScreen from "./GameScreen";
import LeaderboardScreen from "./LeaderboardScreen";
import SignUpScreen from "./SignUpScreen";

function LoginScreen() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const navigation = useNavigation();

  const login = async () => {
    try {
      const user = AsyncStorage.getItem(username);
      const userObJ = AsyncStorage.getItem(JSON.parse(user));
      if (user !== null && userObJ.password === password) {
        setUsername("");
        setPassword("");
        navigation.navigate(GameScreen);
      } else {
        Alert.alert("Invalid input", "Invalid username or password");
      }
    } catch (error) {
      Alert.alert("Invalid input", "Invalid username or password");
    }
  };

  const signUp = () => {
    navigation.navigate(SignUpScreen);
  };

  const leaderboard = () => {
    navigation.navigate(LeaderboardScreen);
  };

  const usernameInputHandler = (enteredText) => {
    setUsername(enteredText);
  };

  const passwordInputHandler = (enteredText) => {
    setPassword(enteredText);
  };

  return (
    <View style={styles.rootContainer}>
      <MainLogo cStyle={styles.logo} />
      <Text style={styles.text}>Login to play</Text>
      <LoginForm
        usernameChange={usernameInputHandler}
        passwordChange={passwordInputHandler}
      />
      <LoginButtons login={login} signUp={signUp} leaderboard={leaderboard} />
    </View>
  );
}

export default LoginScreen;

const styles = StyleSheet.create({
  rootContainer: {
    flex: 1,
    backgroundColor: Colors.color256,
    alignItems: "center",
  },
  text: {
    color: Colors.darkBrown,
    fontSize: Fonts.big,
    fontWeight: "bold",
  },
});
