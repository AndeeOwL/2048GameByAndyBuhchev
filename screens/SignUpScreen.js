import { useNavigation } from "@react-navigation/native";
import { useState } from "react";
import { StyleSheet, Text, View } from "react-native";
import { Colors } from "../constants/Colors";
import { Fonts } from "../constants/Fonts";
import Logo from "../components/Logo";
import SignUpForm from "../components/SignUp/SignUpForm";
import LoginScreen from "./LoginScreen";
import Button from "../components/Button";
import { registerUser, validSignUpCredentials } from "../services/userService";

function SignUpScreen() {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");

  const navigation = useNavigation();

  const register = async () => {
    const isValid = validSignUpCredentials(username, password, confirmPassword);
    if (isValid) {
      const isRegistered = await registerUser(username, password);
      console.log(isRegistered);
      if (isRegistered) {
        navigation.navigate(LoginScreen);
      } else return;
    }
  };

  const usernameInputHandler = (enteredText) => {
    setUsername(enteredText);
  };

  const passwordInputHandler = (enteredText) => {
    setPassword(enteredText);
  };

  const confirmPasswordInputHandler = (enteredText) => {
    setConfirmPassword(enteredText);
  };

  return (
    <View style={styles.rootContainer}>
      <Logo logoContainer={styles.logo} logoText={styles.logoText} />
      <Text style={styles.text}>Register</Text>
      <SignUpForm
        usernameChange={usernameInputHandler}
        passwordChange={passwordInputHandler}
        confirmPasswordChange={confirmPasswordInputHandler}
      />
      <Button
        textStyle={styles.buttonText}
        onPress={register}
        value={"Register"}
      />
    </View>
  );
}

export default SignUpScreen;

const styles = StyleSheet.create({
  rootContainer: {
    flex: 1,
    backgroundColor: Colors.color256,
    alignItems: "center",
  },
  text: {
    color: Colors.darkBrown,
    fontSize: Fonts.big,
    fontWeight: "bold",
  },
  buttonText: {
    marginHorizontal: 15,
    padding: 15,
    backgroundColor: Colors.color8,
    fontSize: Fonts.medium,
    fontWeight: "bold",
    margin: 15,
  },
  logo: {
    marginTop: 50,
    marginBottom: 50,
    backgroundColor: Colors.color1024,
    width: 150,
    height: 150,
    justifyContent: "center",
    alignItems: "center",
  },
  logoText: {
    fontSize: Fonts.extraBig,
    fontWeight: "bold",
  },
});
