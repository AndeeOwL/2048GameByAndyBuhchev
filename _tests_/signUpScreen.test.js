import { render } from "@testing-library/react-native";
import { fireEvent } from "@testing-library/react-native/build";
import LoginScreen from "../screens/LoginScreen";
import SignUpScreen from "../screens/SignUpScreen";

const mockedNavigate = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe("Tests SignUpScreen component", () => {
  test("renders correctly", () => {
    const tree = render(<SignUpScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("it should render right amount of elements", () => {
    const { getAllByText, getAllByPlaceholderText } = render(<SignUpScreen />);
    const logo = getAllByText("2048");
    expect(logo).toHaveLength(1);
    const text = getAllByText("Register");
    expect(text).toHaveLength(2);
    const username = getAllByPlaceholderText("username");
    expect(username).toHaveLength(1);
    const password = getAllByPlaceholderText("password");
    expect(password).toHaveLength(1);
    const confirm = getAllByPlaceholderText("confirm password");
    expect(confirm).toHaveLength(1);
  });

  test("it should change username and password and confirm password", () => {
    const { getByPlaceholderText } = render(<SignUpScreen />);
    const username = getByPlaceholderText("username");
    const password = getByPlaceholderText("password");
    const confirm = getByPlaceholderText("confirm password");
    fireEvent.changeText(username, "andrean");
    fireEvent.changeText(password, "12345678");
    fireEvent.changeText(confirm, "12345678");
  });

  test("Pressed register with invalid inputs should not navigate", () => {
    const { getByPlaceholderText, getAllByText } = render(<SignUpScreen />);
    const username = getByPlaceholderText("username");
    const password = getByPlaceholderText("password");
    const confirm = getByPlaceholderText("confirm password");
    fireEvent.changeText(username, "andrean");
    fireEvent.changeText(password, "12345678");
    fireEvent.changeText(confirm, "12345678");
    const register = getAllByText("Register");
    fireEvent.press(register[1]);
    expect(mockedNavigate).toHaveBeenCalledTimes(0);
  });

  test("Pressed register with valid inputs should navigate", () => {
    const { getAllByText, getByPlaceholderText } = render(<SignUpScreen />);
    const register = getAllByText("Register");
    const username = getByPlaceholderText("username");
    const password = getByPlaceholderText("password");
    const confirm = getByPlaceholderText("confirm password");
    fireEvent.changeText(username, "andrean");
    fireEvent.changeText(password, "12345678");
    fireEvent.changeText(confirm, "12345678");
    fireEvent.press(register[1]);
    expect(mockedNavigate).toHaveBeenCalledWith(LoginScreen);
  });
});
