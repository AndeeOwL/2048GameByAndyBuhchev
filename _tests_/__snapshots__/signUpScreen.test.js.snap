// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests SignUpScreen component renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fae873",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5db31",
        "height": 150,
        "justifyContent": "center",
        "marginBottom": 50,
        "marginTop": 50,
        "width": 150,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 35,
          "fontWeight": "bold",
        }
      }
    >
      2048
    </Text>
  </View>
  <Text
    style={
      Object {
        "color": "#4d4646",
        "fontSize": 32,
        "fontWeight": "bold",
      }
    }
  >
    Register
  </Text>
  <View>
    <TextInput
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="username"
      style={
        Object {
          "backgroundColor": "#e0dbc1",
          "fontSize": 26,
          "height": 50,
          "margin": 15,
          "padding": 10,
          "width": 300,
        }
      }
    />
    <TextInput
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="password"
      style={
        Object {
          "backgroundColor": "#e0dbc1",
          "fontSize": 26,
          "height": 50,
          "margin": 15,
          "padding": 10,
          "width": 300,
        }
      }
    />
    <TextInput
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="confirm password"
      style={
        Object {
          "backgroundColor": "#e0dbc1",
          "fontSize": 26,
          "height": 50,
          "margin": 15,
          "padding": 10,
          "width": 300,
        }
      }
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      onPress={[Function]}
      style={
        Object {
          "backgroundColor": "#f5a164",
          "fontSize": 26,
          "fontWeight": "bold",
          "margin": 15,
          "marginHorizontal": 15,
          "padding": 15,
        }
      }
    >
      Register
    </Text>
  </View>
</View>
`;
